### PR TITLE
Support saving files from test to report

### PR DESCRIPTION
**Pull request by sveinungkb**
This will need some polish probably, but working concept of how files can be saved from any test with.
Consider it a spark for discussion or contribution.
Required some refactoring to how files are pulled to support image + files folder.

`Spoon.save(getInstrumentation().getTargetContext(), someFile);`

And then attached to the report:

### CODE REVIEW

**Comment by edenman** [comment]
@sveinungkb I don't entirely follow.  What's the usecase for this?

**Comment by sveinungkb** [comment]
For us specifically this use case came up for saving a log file (think a http request log) with the test result so we can verify network traffic related to the screenshots grabbed in the test run.
Some other similar use cases would be saving the `SharedPreferences` files or a sqlite used so a test failure can be debugged.

File: spoon-client/src/main/java/com/squareup/spoon/Spoon.java [hunk 1]
```diff
@@ -130,23 +136,91 @@ private static void drawDecorViewToBitmap(Activity activity, Bitmap bitmap) {
...
+    chmodPlusR(target);
+
+    final BufferedInputStream is = new BufferedInputStream(new FileInputStream(source));
+    final BufferedOutputStream os = new BufferedOutputStream(new FileOutputStream(target));
+    byte [] buffer = new byte[4096];
+    while (is.read(buffer) > 0) {
+      os.write(buffer);
+    }
+    is.close();
+    os.close();
```
Comment (JakeWharton) [inline spoon-client/src/main/java/com/squareup/spoon/Spoon.java]:
```java
Files.copy(source, target)
```

Comment (JakeWharton) [inline spoon-client/src/main/java/com/squareup/spoon/Spoon.java]: Whoops this is the client. No Guava!

File: spoon-client/src/main/java/com/squareup/spoon/Spoon.java [hunk 2]
```diff
@@ -8,11 +8,16 @@
 import android.os.Looper;
 import android.util.Log;
 import android.view.View;
+
```
Comment (edenman) [inline spoon-client/src/main/java/com/squareup/spoon/Spoon.java]: kill this line (installing https://github.com/square/java-code-styles should do this for you)

File: spoon-client/src/main/java/com/squareup/spoon/Spoon.java [hunk 3]
```diff
@@ -130,23 +136,91 @@ private static void drawDecorViewToBitmap(Activity activity, Bitmap bitmap) {
...
+    final BufferedOutputStream os = new BufferedOutputStream(new FileOutputStream(target));
+    byte [] buffer = new byte[4096];
+    while (is.read(buffer) > 0) {
+      os.write(buffer);
+    }
+    is.close();
+    os.close();
+  }
+
+  private static File filesDirectory(Context context,
```
Comment (edenman) [inline spoon-client/src/main/java/com/squareup/spoon/Spoon.java]:
```
  private static File filesDirectory(Context context, String directoryType, String testClassName,
      String testMethodName) throws IllegalAccessException {
```

File: spoon-runner/src/main/java/com/squareup/spoon/DeviceTestResult.java [hunk 4]
```diff
@@ -128,7 +142,9 @@ public DeviceTestResult build() {
       if (log == null) {
         log = Collections.emptyList();
       }
-      return new DeviceTestResult(status, exception, duration, screenshots, animatedGif, log);
+      return new DeviceTestResult(status, exception, duration,
+              screenshots, animatedGif, log, files);
     }
+
```
Comment (edenman) [inline spoon-runner/src/main/java/com/squareup/spoon/DeviceTestResult.java]: remove this line

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java [hunk 5]
```diff
@@ -146,7 +152,9 @@ private void printStream(InputStream stream, String tag) throws IOException {
     }
   }
 
-  /** Execute instrumentation on the target device and return a result summary. */
+  /**
```
Comment (edenman) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java]: revert this change

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java [hunk 6]
```diff
@@ -235,60 +243,21 @@ public DeviceResult run(AndroidDebugBridge adb) {
...
-            }
-          }
-        }
+        handleImages(result, screenshotDir);
         FileUtils.deleteDirectory(screenshotDir);
       }
+      if (testFilesDir.exists()) {
+        fileDir.mkdirs();
+        handleFiles(result, testFilesDir);
+//     FileUtils.deleteDirectory(testFilesDir);
```
Comment (edenman) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java]: you'll need to add this back in, yeah?

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java [hunk 7]
```diff
@@ -327,14 +384,15 @@ private void adbPull(IDevice device, FileEntry remoteDirName, String localDirNam
...
+  private FileEntry getScreenshotDirOnInternalStorage(final String dir) {
     String appPackage = instrumentationInfo.getApplicationPackage();
-    String internalPath = "/data/data/" + appPackage + "/" + SCREENSHOT_DIR;
+    String internalPath = "/data/data/" + appPackage + "/" + dir;
     return obtainDirectoryFileEntry(internalPath);
   }
 
-  private static FileEntry getScreenshotDirOnExternalStorage(IDevice device) throws Exception {
-    String externalPath = getExternalStoragePath(device) + "/" + SCREENSHOT_DIR;
+  private static FileEntry getScreenshotDirOnExternalStorage(IDevice device,
```
Comment (edenman) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java]:
```
  private static FileEntry getScreenshotDirOnExternalStorage(IDevice device, final String dir)
      throws Exception {
```

File: spoon-runner/src/main/java/com/squareup/spoon/html/HtmlDevice.java [hunk 8]
```diff
@@ -74,6 +74,8 @@ static HtmlDevice from(String serial, DeviceResult result, File output) {
   }
 
   static final class TestResult implements Comparable<TestResult> {
+
```
Comment (edenman) [inline spoon-runner/src/main/java/com/squareup/spoon/html/HtmlDevice.java]: drop these two lines

File: spoon-runner/src/main/java/com/squareup/spoon/html/HtmlDevice.java [hunk 9]
```diff
@@ -100,13 +106,15 @@ static TestResult from(String serial, DeviceTest test, DeviceTestResult result,
...
+    public final List<HtmlUtils.SavedFile> files;
+    public final boolean hasFiles;
     public final String animatedGif;
     public final HtmlUtils.ExceptionInfo exception;
 
     TestResult(String serial, String className, String methodName, String classSimpleName,
-        String prettyMethodName, String testId, String status,
-        List<HtmlUtils.Screenshot> screenshots, String animatedGif,
-        HtmlUtils.ExceptionInfo exception) {
+               String prettyMethodName, String testId, String status,
```
Comment (edenman) [inline spoon-runner/src/main/java/com/squareup/spoon/html/HtmlDevice.java]: revert these formatting changes

**Comment by edenman** [comment]
Mostly looks good, just some style stuff to clean up.

**Comment by edenman** [comment]
Also: can you add something to the readme that has an example of how to use this?

**Comment by sveinungkb** [comment]
@edenman @JakeWharton thanks for your time, I addressed your PR feedback and added a readme section. Please give it another pass.

File: spoon-runner/src/main/java/com/squareup/spoon/DeviceTestResult.java (hunk 4 above)
Comment (edenman) [inline spoon-runner/src/main/java/com/squareup/spoon/DeviceTestResult.java]: bump

File: spoon-runner/src/main/java/com/squareup/spoon/html/HtmlDevice.java [hunk 10]
```diff
@@ -100,13 +104,15 @@ static TestResult from(String serial, DeviceTest test, DeviceTestResult result,
...
+    public final List<HtmlUtils.SavedFile> files;
+    public final boolean hasFiles;
     public final String animatedGif;
     public final HtmlUtils.ExceptionInfo exception;
 
     TestResult(String serial, String className, String methodName, String classSimpleName,
-        String prettyMethodName, String testId, String status,
-        List<HtmlUtils.Screenshot> screenshots, String animatedGif,
-        HtmlUtils.ExceptionInfo exception) {
+                       String prettyMethodName, String testId, String status,
```
Comment (edenman) [inline spoon-runner/src/main/java/com/squareup/spoon/html/HtmlDevice.java]: revert this formatting

**Comment by sveinungkb** [comment]
Blocked on anything from me on this one? Would be great for us to move to a public version instead of our private snapshot.

**Comment by edenman** [comment]
@sveinungkb sorry, didn't see the last push.  Two things:

- Squash the commits.
- Fill out our CLA: https://docs.google.com/a/squareup.com/forms/d/13WR8m5uZ2nAkJH41k7GdVBXAAbzDk00vxtEYjd6Imzg/viewform?formkey=dDViT2xzUHAwRkI3X3k5Z0lQM091OGc6MQ

**Comment by sveinungkb** [comment]
:white_check_mark: _Squash the commits_
:white_check_mark: _Fill out our CLA_

**Comment by edenman** [comment]
Thanks for the contribution, @sveinungkb!  I'll try to make a release later today.

**Comment by edenman** [comment]
Where by later today I mean now: 1.2.0 just pushed to Maven Central.  Should be live in a couple hours.

**Comment by JakeWharton** [comment]
https://github.com/square/spoon/blob/master/CHANGELOG.md is waaay out of date. README still says 1.1.9.

Also it goes live in 10m or less, so it's already up! http://repo1.maven.org/maven2/com/squareup/spoon/spoon-runner/1.2.0/

**Comment by JakeWharton** [comment]
Oh wait it's only one version out of date. Herp derp numbers.

**Comment by edenman** [comment]
(updated)

**Comment by sveinungkb** [comment]
Thanks guys!
